skillmap/src/App.tsx

```diff
@@ -222,72 +226,62 @@ class AppImpl extends React.Component<AppProps, AppState> {
                     ...currentUser,
                     mapProgress: {}
                 }
 
-                for (const url of urls) {
+                // Patch all of the header ids in the user state and copy
+                // over the local progress that doesn't exist in the signed in
+                // user
+                const localUrls = Object.keys(localUser.mapProgress);
+                for (const url of localUrls) {
                     newUser.mapProgress[url] = {
-                        ...currentUser.mapProgress[url],
-                    };
-
-                    if (!localUser.mapProgress[url]) continue;
+                        ...localUser.mapProgress[url]
+                    }
 
                     const maps = Object.keys(localUser.mapProgress[url]);
                     for (const map of maps) {
-                        newUser.mapProgress[url][map] = {
-                            ...currentUser.mapProgress[url][map]
-                        };
-
                         // Only copy over state if the user hasn't started this map yet
-                        if (Object.keys(newUser.mapProgress[url][map].activityState).length !== 0) {
-                            continue;
-                        }
-
-                        const activityState: {[index: string]: ActivityState} = {};
-                        newUser.mapProgress[url][map].activityState = activityState;
-
-                        const signedInProgress = currentUser.mapProgress[url][map].activityState;
-                        const localProgress = localUser.mapProgress[url][map].activityState
-
-                        for (const activity of Object.keys(signedInProgress)) {
-                            const oldState = signedInProgress[activity];
-                            activityState[activity] = {
-                                ...oldState,
-                                headerId: oldState.headerId ? (headerMap[oldState.headerId] || oldState.headerId) : oldState.headerId
+                        if (!hasUrlBeenStarted(currentUser, url)) {
+                            newUser.mapProgress[url][map] = {
+                                ...localUser.mapProgress[url][map]
                             };
+                            const activityState: {[index: string]: ActivityState} = {};
+                            newUser.mapProgress[url][map].activityState = activityState;
+
+                            const localProgress = localUser.mapProgress[url][map].activityState
+                            for (const activity of Object.keys(localProgress)) {
+                                const localActivity = localProgress[activity];
+                                if (localActivity.headerId) {
+                                    activityState[activity] = {
+                                        ...localActivity,
+                                        headerId: headerMap[localActivity.headerId] || localActivity.headerId
```

Comment:
I think...the headerId should theoretically ALWAYS be in the headerMap. But I don't want to bet on it, so I copied a lot of the old logic here for safety...

Reply:
there is an (admittedly very brief) time where this can happen. for example, if they refresh when loading a tutorial